services-bytebuf-codec/src/main/java/io/scalecube/services/transport/api/ServiceMessageCodec.java

```diff
@@ -9,18 +13,102 @@
 import io.scalecube.services.api.ServiceMessage;
 import io.scalecube.services.exceptions.MessageCodecException;
 import java.lang.reflect.Type;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.function.BiFunction;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import reactor.util.annotation.Nullable;
 
 public final class ServiceMessageCodec {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(ServiceMessageCodec.class);
 
   private final HeadersCodec headersCodec;
 
+  private final Map<String, DataCodec> dataCodecs;
+
+  /** Message codec with default Headers/Data Codecs. */
+  public ServiceMessageCodec() {
+    this(null, null);
+  }
+
+  /**
+   * Create instance from headersCodec and set of DataCodec.
+   *
+   * <p>If codecs are not specified by the user ({@code dataCodecs == null}), DataCodecs obtained
+   * through the SPI mechanism are used. If the user sets several DataCodecs for one Content Type,
+   * then the last one specified is used. User's DataCodec always override DataCodecs from SPI.
+   *
+   * <p>Default HeadersCodec is BinaryHeadersCodec. This is lightweight binary codec written on
+   * vanilla java.
+   *
+   * @param headersCodec codec for message headers. Default, {@link BinaryHeadersCodec}
+   * @param dataCodecs codecs for message body. Codec will select by Message Content Type.
+   */
+  public ServiceMessageCodec(
+      @Nullable HeadersCodec headersCodec, @Nullable Collection<DataCodec> dataCodecs) {
+    this.headersCodec = headersCodec == null ? new BinaryHeadersCodec() : headersCodec;
+    Map<String, DataCodec> defaultCodecs = DataCodec.INSTANCES;
+    if (dataCodecs == null) {
+      this.dataCodecs = defaultCodecs;
+    } else {
+      this.dataCodecs =
+          dataCodecs.stream()
+              .collect(
+                  collectingAndThen(
+                      toMap(DataCodec::contentType, identity(), (c1, c2) -> c2),
+                      usersCodec -> {
+                        Map<String, DataCodec> buffer = new HashMap<>(defaultCodecs);
```

Comment:
IMHO this map is redundant.

Reply:
@segabriel 
`defaultCodecs` are being overwritten with client provided one.

Reply:
defaultCodec maybe is immutable map